src/ast/irbuilderbpf.cpp

```diff
@@ -830,6 +830,71 @@ CallInst *IRBuilderBPF::CreateGetNs(bool boot_time, const location &loc)
   return CreateHelperCall(fn, gettime_func_type, {}, "get_ns", &loc);
 }
 
+Value *IRBuilderBPF::CreateIntegerArrayCmp(Value *ctx,
+                                           Value *val1,
+                                           Value *val2,
+                                           const SizedType &elem_type,
+                                           const uint64_t n,
+                                           const bool inverse,
+                                           const location &loc)
+{
+  /*
+   // This function compares each character of the two arrays.
+   // It returns true if all are equal and false if any are different
+   // cmp([]char val1, []char val2)
+   {
+      for (size_t i = 0; i < n; i++)
+      {
+        if (val1[i] != val2[i])
+        {
+          return false;
+        }
+      }
+      return true;
+   }
+*/
+  size_t elem_size = elem_type.GetSize();
+  Value *offset, *l_addr, *r_addr, *ll, *rr;
+  AllocaInst *l = CreateAllocaBPF(elem_type, "ll");
+  AllocaInst *r = CreateAllocaBPF(elem_type, "rr");
+
+  Function *parent = GetInsertBlock()->getParent();
+  AllocaInst *store = CreateAllocaBPF(getInt1Ty(), "arraycmp.result");
+  BasicBlock *str_ne = BasicBlock::Create(module_.getContext(),
+                                          "arraycmp.false",
+                                          parent);
+  BasicBlock *done = BasicBlock::Create(module_.getContext(),
+                                        "arraycmp.done",
+                                        parent);
+
+  CreateStore(getInt1(!inverse), store);
+  for (size_t i = 0; i < n; i++)
+  {
+    BasicBlock *char_eq = BasicBlock::Create(module_.getContext(),
+                                             "arraycmp.loop",
+                                             parent);
+    offset = CreateMul(getInt64(i), getInt64(elem_size));
+    l_addr = CreateAdd(val1, offset);
+    r_addr = CreateAdd(val2, offset);
+    CreateProbeRead(ctx, l, getInt32(1), l_addr, AddrSpace::kernel, loc);
```

Comment:
If the value is on stack already (e.g. when using `pton`) then you should use GEP instead of proberead.

Reply:
@viktormalik Thanks! So the proper way is, use a GEP when the value is on stack, use a proberead when it is not. Am I understanding it correctly? How should determine if the value is on stack, use IsCtxAccess?

Reply:
Sorry for a late answer. There's a function `shouldBeOnStackAlready` which should be used to detect this (although you'll have to add `pton` there for your usecase).

Reply:
@viktormalik Cool, will try that. Besides, may i know what does the CtxAccess mean?

Reply:
Context is a special structure that is passed to each BPF program. The exact information passed within the context depends on the program type.

Reply:
Thanks. I found `shouldBeOnStackAlready` only recognizes those specific `SizedType`, while `pton` is returning byte arrays only. May I use `is_internal` instead?

Reply:
Yeah, `is_internal` makes sense. And you can use `onStack` instead of `shouldBeOnStackAlready`, then.

Reply:
@viktormalik Great, now the code is using `is_internal`. Please let me if any other issues. Thanks!